### ISSUE TITLE
[Feature request] Synchronization of plugins and their settings. 

### ISSUE THREAD

**Issue by gitmaker032** [vrtmrz/obsidian-livesync#16]
I would like to suggest a killer feature that would greatly increase the usefulness of the plugin:
The ability to synchronize plugins (core/community) and their settings between devices. This is a VERY useful thing if you have several devices/vaults, and even more so if several people are using the vault. Now you have to configure the plugins on each device by hand. Set the same settings. This is not very convenient especially if the list of plugins/their settings change often.  

Required features: 
1. Showing the plugins available in the remote DB and allowing you to choose which of them you want to synchronize with local vault. Options "Synchronize all plugins" / "Synchronize selected plugins".
2. If a plugin is not installed in your local vault, show "This plugin not installed. Copy the plugin and apply settings/data from remote DB?". 
3. If the plugin already has local settings (in other words, the plugin was previously installed in the local vault), suggest "Overwrite existing local settings/data of the plugin?".
4. Optional. When manipulating, show the version of the plugins (which one is installed and which one is in the "cloud") and warn that the version will be upgraded/downgraded.
5. Optional. Ability to choose: synchronize plugins (all / only selected) at each replication or manually (by default). 

I hope Obsidian does not disallow a particular plugin to manipulate the directories of other plugins. 

P.S: Thanks again for the quick and useful updates of the plugin.

**Comment by vrtmrz** [comment]
Thank you for giving me the idea!
I read some articles and other plugins, and found some issue around it.
1. Some people are using BRAT for testing.
2. Automatically sync and or selective sync perhaps make some troubles between desktop and mobile.
3. Some plugin have some setting value that shouldn't be modified except by their Settings dialog (e.g., In this plugin, the database will be corrupted if passphrase and encrypt modified from the others.)

So, it will be dangerous option.

But I inspired, if we can dump the configuration of plugins into markdown file, or applying from markdown file, it could be very useful, and synchronizable by this plugin.

I don't sure it can, but I'll try this as new plugin.

And I'm very happy to hear that my plugin getting be useful, Thank you for everything.

**Comment by gitmaker032** [comment]
I understand the danger. But that's what a lot of people are doing now. They take the vault folder and copy it to a new device (including mobile). Then they just open it. If the plugin does not support the mobile version, it is automatically disabled. In all other cases, the settings that were on the PC are applied. So far I haven't had any problems with this. Plus, if there is a problem with some plugin it is possible not to synchronize it, for example. The plugin is supposed to copy the plugin folders and replace it with a version from the remote DB. For most plugins I think this should work fine.

Maybe this can be made a separate part of the existing plugin. Mark it as a danger zone. I would just say it all relates to synchronization. The point is to sync not only the vault data, but also the plugins/settings. After all, it often determines exactly how the content will be displayed, etc. Some plugins have a large number of settings, it is inconvenient on each device to edit them separately and make sure they are identical. 

I hope it can be implemented in one form or another. I will gladly take part in testing :)

**Comment by vrtmrz** [comment]
I implemented this feature once.
But I forgot cloudant's limitations of document size. (It's 1MB)
Self-hosted LiveSync is bigger than 1MB. lol.
So, if you don't use CouchDB, please wait for a little more.

**Comment by vrtmrz** [comment]
@gitmaker032 
This feature became compatible with IBM Cloudant today!
Could you test this?

**Comment by gitmaker032** [comment]
Thank you for implementing this feature! 🎉

I'm still testing, here's what I've managed to replace so far: 

1. After saving plugins (replication is successful), the list of plugins is not displayed. The Reload button also has no effect on the display. Just appears an empty table. Renaming the device and then saving the plugins helps.

![image](https://user-images.githubusercontent.com/67278783/143778451-6bf56fd2-297c-49e7-9593-79c3745315ac.png)

2. Bad displaying of plugins table on mobile version. I have to use my phone in horizontal mode. It is worth to change the layout.

![image](https://user-images.githubusercontent.com/67278783/143778580-5b9be8b8-32e4-48fc-9098-da13a2022169.png)


3. In recent versions of the plugin in the status bar sometimes appear... rockets :) Sometimes there are a lot of them. Really a lot.

![image](https://user-images.githubusercontent.com/67278783/143778493-ad2fbee2-4a94-4049-abca-062401c21793.png)


4. Common to all plugin buttons: 
a) Styles for button states are not applied: pressed, focus, disabled. Because of this, it's hard to tell if the button was clicked, especially in the mobile version.
b) There is no indicator of task completion after pressing a button. For longer tasks, it would be better to change the button style to "Disabled" and the button text to "In progress...". So that it is clear that the button is pressed and the process is in progress. 

5. And one more thing 🍎: maybe the status bar should show the number of megabytes received and sent, rather than pieces of data. I think this is much more useful information. Right now it is just numbers, which show some process, but they are not so meaningful on their own. Alternatively, you can let the user choose how this information is displayed. 

>This feature became compatible with IBM Cloudant today!
>Could you test this?

I use only my own server, I can't test IBM service, I can't even make an account there (instant blocked on registration).

**Comment by vrtmrz** [comment]
Thank you for testing!

1. Yes, this table doesn't show the own information of the device. Only information on other devices has been shown.  
If renamed, saved data are detected as other devices' one and be shown.
I backup old settings as like so.
But it should be option.

2. You're right. The Mobile version is terribly bad... I'd like to make it better.

3. Each rocket means 'I have to check it and if there're any changes, save it later.'. If ten rockets showed, the plugin is pending saving for ten files.
But, actually a lot. I'll make it better, like candles on dad's birthday cake.

4. Button style is actually indistinct. I'll try to make it better.

5. Transferred bytes is maybe uncountable, it is covered up by PouchDB. but it's may useful. I check it.

I'm really happy to hear your impressions and advice! 
After I wrote the docs, I will specification with your advice.

Thank you for letting me good daddy on weekends :)

Wow, you too? My first account is the same as yours. Create then banned. But the second one is still living.

**Comment by vrtmrz** [comment]
Fixed 1, 2 and 3 at ver 0.1.22.

**Comment by vrtmrz** [comment]
Progress:
- [ ] Automatic saving and replication.
- [ ] Notiify new plugin arrived.
- [ ] Notifiy new plugin's setting arrived.

**Comment by gitmaker032** [comment]
Great! Waiting for updates :) 

1. I think there is still not enough button "Synchronize all plugins". Now you have to do it one plugin at a time manually (probably better to use checkboxes and one button than 20 separate buttons "Use"). Which is not very convenient. Also worth adding the "Synchronize only active plugins" toggle. Because the list of plugins on PC and phone can be very different. In this case, it makes no sense to synchronize all plugins at once, but only those that are used on a particular device (those that are active).  

2. For those users who don't need the plugins synchronization feature (or who use it very rarely), it can be useful to make a toggle that completely hides the corresponding settings. For even now the plugin settings page looks quite long.

**Comment by vrtmrz** [comment]
Thank you for waiting! 
I've implemented Automatic saving and deleting of the old device's settings.
And notification is also implemented, but it looks like something is wrong.

Quite a long setting page is also troublesome, I've tried to split it into tab pages.
And more, "Apply all newer plugins" and "Apply all newer settings" are implemented. But it excepts Self-hosted LiveSync. (to avoid to shuts down this plugin while upgrading progress).

**Comment by vrtmrz** [comment]
@gitmaker032 
Sorry for being late! This feature is now in beta!
Could you please try this?

**Comment by vrtmrz** [comment]
You're right, we lack the documentation.
Thank you for testing, and I'm sorry for making you confused about ambiguous UI.

And really looks like something went wrong to sync plugins.
Looks like there are some different encrypting passphrases. but we can't pick and fix it. We need more implementation.

> In many cases I never managed to sync the plugins. After clicking "Apply (newer)", nothing changes. A notification tells me that it was successful (a plugin files updated). But the settings of the plugin do not change. Although before that I have clearly saved the plugins with the corresponding button on another device.

Really thank you! I didn't notice that, as I close the settings dialog immediately. We need the new dialog.
Perhaps we must not modify settings in the setting dialog without using the setting dialog. Displaying settings and real settings looks like being divergent.

**Comment by vrtmrz** [comment]
@gitmaker032 
Hi! After some hard work, a new dialog is implemented!
It popped out from the setting dialog. So we can avoid this problem.
>  But the settings of the plugin do not change. Although before that I have clearly saved the plugins with the corresponding button on another device.

Some limitations are left. When installed a new plugin, It doesn't show up in the dialog till press "Reload plugins" in "Community plugins" in Obsidian's setting.
But it should be much better than before. Could you please try this?

![image](https://user-images.githubusercontent.com/45774780/154240684-7f3f90d4-948a-4e75-b4ea-e57cd803f928.png)

**Comment by pali6** [comment]
This is a lovely addition to this plugin and I thank you for it. Do you think it would be possible to extend this to core plugin settings and maybe settings of Obsidian itself (i.e. `app.json`, `hotkeys.json`, `appearance.json`, etc.)?

**Comment by vrtmrz** [comment]
@pali6 
I'm really sorry to be late to respond. I've tried this as part of this plugin, but it wasn't useful so much.
But I made the utility that picking and putting any arbitrary file.
It would be handy. Could you please try this?

**Comment by vrtmrz** [comment]
@pali6 
> This is a lovely addition to this plugin and I thank you for it. Do you think it would be possible to extend this to core plugin settings and maybe settings of Obsidian itself (i.e. app.json, hotkeys.json, appearance.json, etc.)?

It has been 125 days, and this feature has been implemented; Obsidian's API improvements were also key.
